### PR TITLE
Membership update

### PR DESCRIPTION
**Pull request by 9OP**
## Implements
This PR implements new CLI commands and new RPC for managing cluster membership lifecycle. Among many changes introduce by this PR, the original HTTP JSON API for issuing commands to the cluster has been removed in favor of a single gRPC API and CLI client.

### New RPCs
`ClusterConfiguration`: returns the cluster configuration (peers, leader and timeouts)
`Execute`:  Introduced to replace the HTTP JSON API for interpreting client commands
`Shutdown`: Remotely shutdown a node
`Ping`: Dummy RPC to check if a peer is up and reachable (and not hidden behind a NAT for example).

### Data structure change
Mostly new data structure for the new RPC calls, `ClusterConfiguration`, `NodeConfig`, `LogType`, `ConfigurationUpdateType`.

The most important change is the update of `Peer` which is simplified and contains an `Active` boolean field to indicate if the peer is active (ie. part of the quorum, can become leader).

### Details about cluster membership update
The cluster membership change are not well described in the Raft paper, so I had to improvise a mechanism of my own. It is working and open to improvement and redesign. So far this is how I designed the mechanism to add a new peer to a live cluster:
1. Retrieve the leader addr from the cluster configuration.
2. Send an RPC `Execute` with a `LogConfigurationUpdate` to set the new peer as `Inactive`.
3. Start the new peer (the peer starts on the machine the command is invoked on).
4. Send an RPC `Execute` to set the new peer status to `Active`.

Removing a peer from the cluster is similar, 
1. First retrieve the leader addr from the cluster configuration
2. `Execute` set old peer status as `Inactive`
3. `Execute` remove old from node peers map
4. Eventually, but not required, shutdown old peer.

The rational is the following:
- Reusing the log replication mechanism to duplicate membership update commands accross the cluster.
- Perform any membership update operation in 2 steps:
  - Set peer as active or inactive
  - Add or remove peer

We need to performs membership update operation in 2 steps to prevent the cluster from reaching an inconsistent state, where some node have different `peers` information and thus different `Quorum`. Ensuring that we add a peer first as `inactive` allow to prevent this inconsistent state in case the command fails.

In short, there is no risk to break the cluster if the new peer is added as `Inactive` in a minority of the cluster nodes.

### CODE REVIEW

File: pkg/domain/errors.go [hunk 1]
```diff
@@ -1,27 +1,10 @@
...
-
-func NewNotLeaderError(leader Peer) *NotLeaderError {
-	return &NotLeaderError{leader, &domainError{"not leader"}}
-}
-
-func NewUnknownLeaderError() *UnknownLeaderError {
-	return &UnknownLeaderError{&domainError{"unknown leader"}}
-}
+import "errors"
+
```
Comment (9OP) [inline pkg/domain/errors.go:4]: Cleaner that way, the old way is not idiomatic GoLang error handling.

File: pkg/domain/node.go [hunk 2]
```diff
@@ -245,30 +303,133 @@ func (n *Node) GrantVote(peerId string) {
...
+// 	go (func() { result <- c.evalFsm(query, "QUERY") })()
+// 	return result
+// }
+
+var (
+	errPeerAlreadyExists = errors.New("peer already exists")
+	errPeerDoesNotExist  = errors.New("peer does not exist")
+)
+
+func (n *Node) configurationUpdate(config ConfigurationUpdate) (res ExecuteOutput) {
```
Comment (9OP) [inline pkg/domain/node.go:385]: Cluster configuration use the log replication mechanism to ensure that the changes are duplicated on each node. In my opinion it is nice to reuse the log replication mechanism for managing the cluster state and not only duplicating FSM commands.

File: pkg/domain/peer.go [hunk 3]
```diff
@@ -8,56 +8,43 @@ import (
...
-		P2p string
-		Api string
-	}
-}
-
-func errInvalidAddr(addr string) error {
-	return fmt.Errorf("invalid addr format %s", addr)
+	Id     string
+	Host   netip.AddrPort
+	Active bool
```
Comment (9OP) [inline pkg/domain/peer.go:13]: Mostly simplifying the data structure and adding `Active` which indicates if the peer should be included in the Quorum.

File: pkg/infrastructure/server/cluster.go [hunk 4]
```diff
@@ -1,108 +0,0 @@
-package server
```
Comment (9OP) [inline pkg/infrastructure/server/cluster.go:1]: Farewell sweet JSON API, but you were a burden to maintain along the gRPC API, hope to see you again soon

File: pkg/membership.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+	switch err {
+	case nil:
+		break
+	case errLeaderNotFound:
+		leaderNotFound = true
+	default:
+		return nil, err
+	}
+
+	if leaderNotFound {
```
Comment (9OP) [inline pkg/membership.go:26]: This is a bit tricky, but  put simply:
since we are not able to get the Leader from the cluster configuration (either partition or election) we have a potentially stale configuration.

We can still try to start the node if the node host/id is contained in the configuration and active, because it would mean that the node is already known by the cluster. But not necessarily active.

This is better than nothing, but we should make clear that is is not ideal and the client should by himself that the node is indeed active in the entire cluster and not only on the node from which the configuration was fetched.
